py/picca/pk1d/postproc_pk1d.py

```diff
@@ -862,12 +818,370 @@ def compute_average_pk_redshift(
+
+        func = partial(
+            compute_cov,
+            weight_method,
+            mean_p1d_table,
+            nbins_k,
+        )
+        if number_worker == 1:
+            output_cov = [func(*p) for p in p1d_groups_bootstrap]
+        else:
+            with Pool(number_worker) as pool:
+                output_cov = pool.starmap(func, p1d_groups_bootstrap)
+
+        for izbin in range(nbins_z):
+            boot_cov = []
+            for iboot in range(number_bootstrap):
+                covariance_array = (output_cov[izbin * number_bootstrap + iboot],)
+                boot_cov.append(covariance_array)
+
+            i_min = izbin * nbins_k * nbins_k
+            i_max = (izbin + 1) * nbins_k * nbins_k
+            cov_table["boot_covariance"][i_min:i_max] = np.mean(boot_cov, axis=0)
+            cov_table["error_boot_covariance"][i_min:i_max] = np.std(boot_cov, axis=0)
+
+
+
+
+def compute_p1d_groups(
+    weight_method,
+    nbins_k,
+    zbin_edges,
+    izbin,
+    p1d_table,
+    k_index,
+    snrfit_table,
+    number_worker,
+):
+    """Compute the P1D groups before covariance matrix calculation.
+    Put all the P1D in the same k grid.
+
+    Arguments
+    ---------
+    weight_method: str,
+    Method to weight the data.
+
+    nbins_k (int):
+    Number of k bins.
+
+    zbin_edges (array-like):
+    All redshift bins.
+
+    izbin (int):
+    Current redshift bin being considered.
+
+    p1d_table (array-like):
+    Table of 1D power spectra, with columns 'Pk' and 'sub_forest_id'.
+
+    k_index (array-like):
+    Array of indices for k-values, with -1 indicating values outside of the k bins.
+
+    snrfit_table: numpy ndarray,
+    Table containing SNR fit infos
+
+    number_worker: int
+    Number of workers for the parallelization
+
+    Return
+    ------
+    p1d_weights  (array-like):
+    Weights associated with p1d pixels for all subforest, used in the calculation of covariance.
+
+    covariance_weights (array-like):
+    Weights for all subforest used inside the main covariance sum.
+
+    p1d_groups (array-like):
+    Individual p1d pixels grouped in the same wavenumber binning for all subforest
+    """
+
+    select_z = (p1d_table["forest_z"] < zbin_edges[izbin + 1]) & (
+        p1d_table["forest_z"] > zbin_edges[izbin]
+    )
+    p1d_sub_table = Table(
+        [
+            p1d_table["sub_forest_id"][select_z],
+            k_index[select_z],
+            p1d_table["Pk"][select_z],
+        ],
+        names=("sub_forest_id", "k_index", "pk"),
+    )
+    if weight_method == "fit_snr":
+        snrfit_z = snrfit_table[izbin * nbins_k : (izbin + 1) * nbins_k, :]
+        p1d_sub_table["weight"] = 1 / fitfunc_variance_pk1d(
+            p1d_table["forest_snr"][select_z],
+            snrfit_z[k_index[select_z], 2],
+            snrfit_z[k_index[select_z], 3],
+        )
+    else:
+        p1d_sub_table["weight"] = 1
+
+    p1d_sub_table = p1d_sub_table[p1d_sub_table["k_index"] >= 0]
```

Comment:
are we storing negative k_index somewhere? Why is this cut needed?

Reply:
This was a convention for the case when a k bin computed for the p1d does not fall inside the output k binning. I removed those bins as they do not contribute to the covariance.

Reply:
This cut is needed because I made a method to compute the p1d groups in a vectorized way, and -1 change the last k bin, but it should not change any bin

Reply:
I still don't fully get how this line is changing anything. Wouldn't the selector be always true? Or are there nans or -1s in that table for bins that just aren't filled and that you want to remove here (in which case I would test for not having exactly that filler value and not via >=0)?

Reply:
With the implementation right now, this is just to remove p1d pixels which were not associated with any wavenumber bin. 
I made a method previously for which this cut was necessary, but now it is just to reduce a little the input dataset, removing unecessary pixels.
Comment added.